src/analysis.ts

```diff
@@ -36,10 +39,55 @@ export class ValidationError extends Error {
 
 
 export class Analysis {
-  descriptors: DocumentDescriptor[];
+  private _descriptors: DocumentDescriptor[];
+  private _elementsByTagName = new Map<string, ElementDescriptor>();
+  private _elementsByPackageDir = new Map<string, ElementDescriptor[]>();
+  elementPaths = new Map<ElementDescriptor, string>();
+  locationOffsets = new Map<ElementDescriptor, LocationOffset>();
 
   constructor(descriptors: DocumentDescriptor[]) {
-    this.descriptors = descriptors;
+    this._descriptors = descriptors;
+    const packageGatherer = new PackageGatherer();
+    const elementsGatherer = new ElementGatherer();
+    new AnalysisWalker(this._descriptors).walk([
+      packageGatherer, elementsGatherer
+    ]);
+
+    const elements = elementsGatherer.elementDescriptors;
+
+    for (const element of elements) {
+      if (element.is) {
+        this._elementsByTagName.set(element.is);
+      }
+      let elementPath = elementsGatherer.elementPaths.get(element);
```

Comment:
can you add some comments about what this look is doing? I'm finding the code not super self-documenting